src/Castle.ts

```diff
@@ -0,0 +1,357 @@
+import fetch from 'node-fetch';
+import { reduce } from 'lodash';
+import { IncomingHttpHeaders } from 'http2';
+import AbortController from 'abort-controller';
+import packageJson from '../package.json';
+import pino from 'pino';
+
+const defaultApiUrl = 'https://api.castle.io';
+
+type ActionParameters = {
+  event: string;
+  user_id: string;
+  user_traits?: object;
+  properties?: object;
+  created_at?: string;
+  device_token?: string;
+  context: {
+    ip: string;
+    client_id: string;
+    headers: IncomingHttpHeaders;
+  };
+};
+
+type ActionType = 'allow' | 'deny' | 'challenge';
+
+type FailoverStrategyType = ActionType | 'none';
+
+type AuthenticateResult = {
+  action: ActionType;
+  user_id?: string;
+  device_token?: string;
+  failover?: boolean;
+  failover_reason?: string;
+};
+
+type LoggingParameters = {
+  requestUrl: string;
+  requestOptions: any;
+  response?: Response;
+  err?: Error;
+  body?: any;
+};
+
+// The body on the request is a stream and can only be
+// read once, by default. This is a workaround so that the
+// logging functions can read the body independently
+// of the handlers.
+const getBody = async (response: any) => {
+  if (response.cachedBody) {
+    return response.cachedBody;
+  }
+
+  try {
+    response.cachedBody = await response.json();
+  } catch (e) {
+    if (e.message === 'invalid json response body') {
```

Comment:
what if the message is different? Won't it silence the error?

Reply:
You're right! I should handle all errors the same way here, by setting empty body, I don't have any other use case. Thanks! Fixed

Reply:
Actually even refactored it away